### PR TITLE
feat(0.10.0): consolidated find + check tools (#128)

### LINKED ISSUE #128: epic: consolidate find-usages + type-check clusters into intent-named `find` / `check` (kill agent rg-fallback)
# Consolidate find-usages + type-check clusters into `find` and `check`

**Labels:** epic, tool-consolidation, agent-ux
**Tracking:** part of #100 (fslangmcp agent-UX catch-all)

## Problem (live metrics, v0.9.3)
- **34 tools** registered (Program.fs:228-604) — the surface is the #1 source of agent friction.
- **~6k-token description tax** read on every session before the first call.
- **22 overlapping-intent pairs** force a per-task which-tool decision. The two worst clusters:
  - **find-usages (7 tools):** fcs_find_symbol, fcs_record_field_audit, fcs_find_member_usages, workspace_symbol, fcs_project_symbol_uses, textDocument_references, textDocument_definition.
  - **type-check (5 tools):** workspace_diagnostics, fsharp_compile, fcs_check_file, fcs_parse_and_check_file, fcs_validate_snippet.

Mined-corpus evidence of the cost (40-task developer-advocate audit):
- `fcs_find_symbol` returned `referenceCount:2` for **TraderRole** (missed all ~28 `{ Propose = fun ... }` sites) and `matchedUseCount:0` for **RiskDebatorRole** -> agent rg-fell-back. Two subagents independently arrived at `rg` first.
- `workspace_symbol "Foreground"` found the type but not `style.Foreground` call-sites; `fcs_find_symbol` **requires `path`** (Types.fs:179) which "isn't obvious from the tool name" -> `InvalidArgs: Parameter path`.
- `workspace_diagnostics` came back `{}` but didn't re-check cross-file resolution (FactUnlessWindowsAttribute) -> dotnet build. "Could not tell whether `diagnosticsFileCount:0` meant clean or not-yet-run."
- `fcs_validate_snippet` wants `content`, `fcs_parse_and_check_file` wants `path`, `set_project` wants `projectPath` -> "первый вызов — угадайка"; threw `InfraFailure` on .slnx.

## Proposal
Two intent-named tools; the FCS/FSAC seam hides behind an internal dispatcher (engine surfaced only as a debug `via`):
- **`find(query, ...)`** — "where is this defined/used". `query` is the SOLE required arg; auto-unions defs + refs + record-field sites + member sites.
- **`check(...)`** — "does this compile / did my edit break anything". No required args; returns ONE actionable `verdict`.

Full DECISION-READY spec (signatures, param tables, response shapes, fold map, integration points file:line) attached in the linked design doc.

## Red-team fixes folded in (not waved away)
1. **RELOCATED-CHOICE:** drop the agent-facing `fresh` boolean + actionable groundTruth/staleAcrossFiles. `check()` returns a single `verdict` (clean|errors|unknown); freshness/escalation is internal; cached snapshot is an explicit `speed:"fast"` opt-in only.
2. **DEFAULT-GAP:** the FCS substrate `FindSymbol` (FcsBridge.fs:1812-1859, `GetAllUsesOfAllSymbols` per-compilation at :1853) is single-project — so `scope=auto`=solution-wide is **net-new** (multi-project union sweep, resolve .slnx->member .fsproj, de-dup by (file,range)), NOT a forwarder. Self-heal split by substrate; never emit confident `matched=false` from a single-project FCS result (would false-negative RiskDebatorRole).
3. **COLD-START:** set_project calls `ClearCaches()` (Program.fs:288 -> FcsBridge.fs:3242), so the first find/check is always cold; add a background pre-warm in set_project.

## Surface impact
- **34 -> 24 tools** (12 folded into 2; transient +2 aliases during the deprecation window).
- ~6k-token description tax roughly halved; backend `[FSAC]`/`[FCS in-process]` prefixes stripped from all descriptions.

## Rollout (zero-breakage, telemetry-gated)
Stage 0 dispatcher extraction (behavior-identical) -> Stage 1 find/check GA with old tools as deprecated forwarders into the SAME dispatcher -> Stage 2 docs flip -> Stage 3 per-tool removal gated by fleet hit-rate via the #100 UX-report loop (fold tools first, LSP deprecate-aliases last). Outputs are a strict SUPERSET of legacy fields; every stage reverts independently.

## Acceptance (A/B on the 40-task corpus)
- rg-fallback rate **< 10%** (from ~35-45%).
- wrong-tool-call rate **-50%**; tokens-to-first-correct-tool **-30%**.
- HARD GATE: zero confident false-negatives on RiskDebatorRole/MarketDataWarning; check single-verdict trust **>=95%** on the gate tasks; `box u` snippet must NOT return `clean`.

## Touch points
Program.fs:498/260 (register), :457/469/493/501 + :262/365/373/433 (re-point handlers), description-prefix strip across :230-560; FcsBridge.fs:1812-1859 (wrap FindSymbol), :3242 (ClearCaches/pre-warm); LspBridge.fs:418/928/942 (FSAC self-heal); Types.fs new FindArgs/CheckArgs (path optional, `content`->`snippet` :313); new Dispatcher.fs; ToolDescriptionSchemaTests + new DispatcherTests (mechanical-equivalence gate).

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Fold map — all 12 cluster tools + adjacency

| Old tool | Backend | Disposition | Reachable as | Reason |
|---|---|---|---|---|
| **fcs_find_symbol** | FCS | **fold** | `find` auto | Substrate of find's default (grouped defs+refs+source ctx). Consolidation promotes `symbolQuery`->`query` and makes the REQUIRED `path` (Types.fs:179) OPTIONAL — the single biggest friction keeping agents on rg. Forwarding alias, telemetry-gated removal. |
| **fcs_record_field_audit** | FCS | **fold** | `find kind=field` + auto-union | Auto-unioned into find default when query resolves to a record type. Closes the most-recurring refactor collision (TraderRole.Propose ~28 `{ Propose = fun ... }` sites missed by fcs_find_symbol/textDocument_references). |
| **fcs_find_member_usages** | FCS | **fold** | `find kind=members` + auto-sweep | Auto-swept when query resolves to a member. Closes Style.Foreground member-accessor rename (workspace_symbol found type, not dotted `x.Foreground` sites). |
| **workspace_symbol** | FSAC | **fold** | `find kind=symbol scope=workspace` | Fast FSAC name-fragment lookup preserved as explicit mode, backend hidden. Fold not keep because its IDE-shaped-without-context output was the documented reason agents fell to rg/Read; default find supersedes it. Aliased for cheap-lookup fleet callers (WorkspaceSymbolArgs.query Types.fs:54). |
| **fcs_project_symbol_uses** | FCS | **deprecate-alias** | `find(query, scope="project")` | Pure forwarder: project-wide uses are a strict subset of find default's grouped `references`; contributes NO new mode. Kept as deprecated alias only to avoid breaking existing callers (incl. the praised `exact=true` "exactly 2 sites" buildHost case — preserved via find(exact:true,scope:project)). Removed when telemetry shows idle. |
| **textDocument_references** | FSAC | **deprecate-alias** | `find kind=position` | Exact-position LSP primitive (ReferencesArgs Types.fs:42). Deprecate-alias not fold: editor/LSP automation passes raw 0-based coords and depends on LSP-shaped output byte-for-byte; forwarder preserves the contract. LAST removal stage (slowest-migrating callers). |
| **textDocument_definition** | FSAC | **deprecate-alias** | `find kind=definition` | Exact-position LSP primitive (PositionArgs Types.fs:32). Considered keep (navigation arguably distinct) but rejected — find already returns definitions; keeping it defeats consolidation. Forwarder preserves LSP-shaped output for IDE automation; last removal stage. |
| **workspace_diagnostics** | FSAC | **fold** | `check(speed="fast")` | Single most-used corpus tool ("single highest-ROI"). Folded — NOT forever-forwarded — because DiagnosticsArgs.fileGlob/severity (Types.fs:60/62) are absorbed into check params, making it a true fold. Legacy semantics preserved byte-for-byte under speed=fast. |
| **fsharp_compile** | FCS | **fold** | `check(scope="project"\|"workspace")` | ParseAndCheckProject path. Corpus rejected it standalone ("duplicates dotnet build without clear advantage") — folding it behind check's internal-escalation verdict removes the standalone-rejection problem (it is no longer a tool the agent picks; it is an engine the dispatcher picks). |
| **fcs_check_file** | FCS | **fold** | `check(scope="file")` (trusted) | Cache-invalidating per-file check (Program.fs:371). Its Program.fs:372 AST-cache caveat becomes check's internal escalation trigger, not a description note. |
| **fcs_parse_and_check_file** | FCS | **fold** | `check(scope="file", speed="fast")` | Per-file parse+check (Program.fs:363). The "MVP, clear green signal, echoes Compile ordering" praise is preserved as check's file-scope payload. |
| **fcs_validate_snippet** | FCS | **fold** | `check(scope="snippet")` | content->snippet rename (Types.fs:313) ends the documented content-vs-code-vs-path guess-and-retry. .slnx member-.fsproj auto-selection + nullness/warn-as-error threading fixed inside check. |

### Explicitly KEPT (adjacent — NOT consolidated; named so the maintainer sees the boundary)
| Tool | Why kept |
|---|---|
| **set_project** | Context setup, upstream of both tools; unchanged except a background pre-warm added (cold-start fix). |
| **fcs_type_at_position / fcs_symbol_at_word** | Type-at-position is a distinct intent. find *borrows* word-anchoring from fcs_symbol_at_word but does not absorb it. |
| **fcs_referenced_symbols / fcs_nuget_types / fcs_nuget_members** | Referenced-assembly/NuGet enumeration — find ROUTES TO these (out-of-scope hint) rather than absorbing; project-local substrate can't serve assembly symbols. |
| **project_health / fsharp_project_inspect / fcs_project_outline / fcs_file_outline / fcs_file_symbols** | Orientation/structure, not find-usages or type-check. |
| **textDocument_completion / formatting / codeAction / rename / fcs_signature_help / fsharp_signature_data / fcs_make_internal_visible / fcs_suggest_open** | Editor IO / repair primitives, out of both clusters. |

## Backward-compatible rollout (zero-breakage, telemetry-gated removal)

**Principle:** new tools and the 12 old tools route into the SAME internal dispatcher (mechanical equivalence). Old tools never change behavior during the window; new-tool outputs are a strict SUPERSET of legacy fields. Removal is gated by fleet hit-rate telemetry via the #100 UX-report loop — NOT by calendar.

### Stage 0 — Dispatcher extraction (1 PR, no surface change)
- Add `Dispatcher.fs` exposing `FindDispatch.run : FindArgs -> Task<JsonNode>` and `CheckDispatch.run : CheckArgs -> Task<JsonNode>`.
- Re-point the 12 existing handler lambdas (Program.fs:457/469/493/501 find-cluster; :261/269/365/373/433 check-cluster) at the dispatcher with thin arg-adapters. Behavior byte-identical; covered by existing FcsBridgeTests/LspBridgeTests. **Ship-gate: full test suite green, descriptions unchanged.**

### Stage 1 — `find` + `check` GA, old tools still primary (1 PR)
- Register `find` (near Program.fs:498) and `check` (near Program.fs:260) via `TypedTool.define<FindArgs>` / `<CheckArgs>`.
- Land the net-new capabilities the substrate lacks: multi-project union sweep (wrap FcsBridge.fs:1812-1859), substrate-split self-heal, snippet nullness threading, check internal-escalation verdict, set_project background pre-warm.
- Strip `[FSAC]`/`[FCS in-process]` prefixes from ALL descriptions (Program.fs:230,238,246,254,262,270,282,329,352,364,372,380,392,404,416,428,440,448,456,464,476,488,500,508,516,528,536,544,552,560) — backend invisible. Update ToolDescriptionSchemaTests to accept find/check under the per-tool token budget.
- Old 12 tools annotated `(deprecated: use find/check)` in their description tail; behavior unchanged. **Surface transiently 34 -> 36.**

### Stage 2 — Docs + advocacy flip (no code)
- AGENT_INTEGRATION.md, README, tools-detailed.md lead with find/check; old tools moved to a "legacy aliases" appendix.
- CLAUDE.md global F# rule updated: canonical entry points become find/check.
- Begin telemetry collection: version-tagged UX reports record per-tool call counts + rg-fallback events on the mined corpus (#100 loop).

### Stage 3 — Data-gated removal (per-tool, staggered)
Remove an old tool ONLY when telemetry shows its fleet call-rate has fallen below threshold for 2 consecutive version tags AND find/check cover its corpus tasks at >=90%. Removal order by migration speed:
1. **Fold tools first** (fcs_find_symbol, fcs_record_field_audit, fcs_find_member_usages, workspace_symbol, fsharp_compile, fcs_check_file, fcs_parse_and_check_file, fcs_validate_snippet, workspace_diagnostics) — superset-compatible, lowest risk.
2. **deprecate-alias tools last** (fcs_project_symbol_uses, then textDocument_references, textDocument_definition) — LSP/editor automation byte-for-byte contract, slowest-migrating callers.
- Each removal is its own PR with a one-version-tag deprecation-warning grace period in the tool description. **Final surface 34 -> 24.**

### Rollback
Any stage reverts independently: Stage 1 register-only PRs can be reverted without touching Stage 0 dispatcher; deleting find/check restores exact prior surface because old handlers still call the dispatcher directly.

### Program.fs — registration & descriptions
- **find registration:** insert `TypedTool.define<FindArgs> "find" <desc> (fun args -> toolResult (runLimited fcsGate (fun () -> FindDispatch.run args)))` adjacent to the existing fcs_find_symbol block (**Program.fs:498-501**).
- **check registration:** insert `TypedTool.define<CheckArgs> "check" ...` adjacent to workspace_diagnostics (**Program.fs:260-261**).
- **Re-point legacy handlers at dispatcher (Stage 0):** find-cluster lambdas at **:457** (ProjectSymbolUses), **:469** (FindMemberUsages), **:493** (RecordFieldAudit), **:501** (FindSymbol); check-cluster at **:262/275** (workspace_diagnostics/CompileProject), **:365** (ParseAndCheckFile), **:373** (CheckFile), **:433** (ValidateSnippet).
- **Strip backend prefixes** `[FSAC]`/`[FCS in-process]` from descriptions at **:230, :238, :246, :254, :262, :270, :282, :329, :352, :364, :372, :380, :392, :404, :416, :428, :440, :448, :456, :464, :476, :488, :500, :508, :516, :528, :536, :544, :552, :560**.
- **set_project background pre-warm (cold-start fix):** at **Program.fs:288** set_project calls `fcsBridge.ClearCaches()` then returns; add a fire-and-forget `Async.Start` that runs ParseAndCheckProject over the active project(s) so the first find/check is warm. Guard with the existing `fcsGate`.

### FcsBridge.fs — find substrate (the net-new sweep lives here)
- **FindSymbol member at :1812-1859** is single-`projectOptions`: `ParseAndCheckProject(projectOptions)` at **:1832**, `projectResults.GetAllUsesOfAllSymbols()` at **:1853** is per-compilation. Wrap in a NEW `member FindSymbolAcrossProjects(args, memberFsprojs)` that: resolves the loaded .sln/.slnx -> member .fsproj list, calls the existing single-project path per project, unions results de-duped by (file,range). Do NOT merely forward.
- Other per-compilation ParseAndCheckProject call-sites that the multi-project sweep must NOT assume cover the solution: **:1096, :1294, :1403, :1564, :2830** (each takes one `projectOptions`).
- **ClearCaches at :3242-3244** (`optionsCache.Clear()` line 119 BoundedCache(10); `projectResultsCache.Clear()` line 120 BoundedCache(3)) — document that this is why post-set_project calls are cold; the pre-warm above is the mitigation.

### LspBridge.fs — FSAC self-heal path
- **:418** `mutable symbolIndexEverWarmed`, reset at **:523**, exposed via **:534 IsSymbolIndexReady**.
- **:928 WorkspaceSymbol** invokes `workspace/symbol` (**:938**); first non-empty sets the warm flag at **:942**. find's FSAC-substrate self-heal (warm-then-retry-once) hooks here. The warmup-window assessment helper is at **:310-321/:369-382** — reuse, do not duplicate.

### Types.fs — new arg records (additive)
- **NEW `FindArgs`** modeled on FcsFindSymbolArgs (**:178**) but with `query` required and **`path` OPTIONAL** (the headline fix; current `path: string` at :179 is mandatory). Add `kind`,`scope`,`exact`,`member`,`field`,`line`,`word`,`occurrence`.
- **NEW `CheckArgs`** unioning FSharpCompileArgs (**:74**, timeoutMs :79), DiagnosticsArgs (**:55-64**, path/fileGlob/severity), FcsParseAndCheckArgs (**:92**), FcsValidateSnippetArgs (**:311**); rename `content` (**:313**) -> `snippet`; add `scope`,`speed`,`verdict` output.
- Legacy records (WorkspaceSymbolArgs :54, ReferencesArgs :42, PositionArgs :32) untouched — their handlers adapt into FindArgs at the dispatcher boundary.

### Dispatcher.fs (NEW) — backend selection table
`find/auto|members|field` -> FCS (FindSymbolAcrossProjects + FindMemberUsages + RecordFieldAudit); `find/kind=symbol|scope=workspace` -> FSAC workspace/symbol; `find/kind=position|definition` -> FSAC textDocument/references|definition. `check/scope=project|workspace + trusted` -> FCS ParseAndCheckProject (CompileProject); `check/scope=file + trusted` -> CheckFile; `check/scope=file + fast` -> ParseAndCheckFile; `check/fast project-wide` -> FSAC publishDiagnostics; `check/scope=snippet` -> ValidateSnippet. Engine surfaced only as `via`.

### Tests
- **tests/FsLangMcp.Tests/ToolDescriptionSchemaTests.fs** — extend the per-tool token-budget allowlist to find/check; assert no `[FSAC]`/`[FCS]` prefix remains.
- **FcsBridgeTests.fs / LspBridgeTests.fs** — add multi-project-sweep union test (RiskDebatorRole-shaped: symbol defined in proj A, used in B/C) and a self-heal-no-false-negative test (assert matched=false ONLY when both substrates empty).
- New **DispatcherTests.fs** — assert each legacy tool and its find/check equivalent produce field-superset-equal payloads (mechanical-equivalence regression gate).

### Harness
Replay the 40-task mined corpus (the EVIDENCE.surface tasks) through subagents in two arms via the #100 version-tagged UX-report loop:
- **Arm A (control):** current 34-tool surface (v0.9.3).
- **Arm B (treatment):** find/check GA build (Stage 1), old tools deprecated-but-present.
Each task run by an independent subagent with identical CLAUDE.md F# tool guidance; subagent emits a structured UX report tagging, per task: tools called (ordered), empty/error responses, whether it fell back to rg/Read/dotnet-build, and tokens consumed before the first call that returned the correct answer. Aggregate across >=3 seeds per task to smooth agent nondeterminism.

### Primary metrics + thresholds (ship gates)
| Metric | Definition | Baseline (Arm A, from corpus) | Ship threshold (Arm B) |
|---|---|---|---|
| **rg-fallback rate** | % of find/check-eligible tasks where the agent fell to rg/Read/dotnet-build despite a semantic tool existing | ~35-45% (collisions list: Foreground, RiskDebatorRole, TraderRole, BrokerError-warmup, MarketDataWarning, FactUnlessWindows, validate-snippet-on-.slnx) | **< 10%** |
| **wrong-tool-call rate** | calls returning empty/error/false-negative that triggered a different tool next (guess-and-retry: content-vs-code-vs-path, workspace_symbol-empty-then-rg) | high on the 12-tool naming-inconsistency + cold-index tasks | **>= 50% reduction** |
| **tokens-to-first-correct-tool** | tokens spent (description reading + failed calls) before the call that returns the right answer | driven by 12 overlapping descriptions + failed first calls | **>= 30% reduction** (12->2 description shrink + bare-call defaults) |

### Hard gates (any failure blocks ship, independent of the above)
- **Zero confident false-negatives** on the two cross-project tasks (RiskDebatorRole used across Stubs/Pipeline/3×Maf*Debator/Program; MarketDataWarning multi-project audit). Any `find` returning `matched=false` for a symbol rg finds is an automatic block — directly tests the red-team DEFAULT-GAP fix.
- **check single-verdict trust >= 95%** on the gate tasks (post-edit "did I break anything", 7- and 17-project runs): % where the agent acted on `verdict` alone without a dotnet-build fallback. Tests the RELOCATED-CHOICE fix.
- **Snippet honesty:** the `box u` under `<Nullable>enable</Nullable>` task must NOT return `verdict:"clean"` (must be `errors` or `unknown` with snippetFlagsMayDiffer), reproducing the FS3261 corpus false-negative under check.
- **Cold-start latency:** first find/check after set_project on the 17-project .slnx completes within the existing fcsGate timeout with pre-warm enabled; record p50/p95 — regression vs Arm A's first workspace_diagnostics call must be <= +1 ParseAndCheckProject worth.

### Secondary (telemetry, informs Stage 3 removal)
Per-old-tool fleet call-rate over consecutive version tags; an old tool is removable when its rate is below threshold for 2 tags AND its corpus tasks are covered by find/check at >=90%.

## Open risks

- MULTI-PROJECT SWEEP COST: scope=auto on a 17-project .slnx runs ParseAndCheckProject per member project on the first cold call (after set_project's ClearCaches at FcsBridge.fs:3242). Background pre-warm mitigates but does not eliminate the O(projects) first-hit latency; if pre-warm hasn't finished, the first find/check still blocks. Needs a real latency measurement on the 17-project corpus before committing scope=auto=solution-wide as the default — may need to cap concurrency or fall back to active-project-only with an explicit 'widen?' hint when N is large.
- INTERNAL dotnet-build ESCALATION IN check: making check silently shell out to `dotnet build` to resolve the FCS AST-cache caveat changes its cost profile from 'cheap type-check' to 'possibly a full build' and introduces a process-spawn dependency the MCP server didn't have on this path. Must be config-gated (allow/deny) and time-boxed; when disallowed, verdict='unknown' is the honest output but reintroduces an agent decision — partial relocation remains. Accepted-risk unless telemetry shows 'unknown' is rare.
- FALSE-NEGATIVE GATE IS HARD TO PROVE EXHAUSTIVELY: the 'never emit confident matched=false from a single-project FCS result' rule is testable on the known RiskDebatorRole/MarketDataWarning shapes, but a symbol used only via reflection, in a project not loaded by the .slnx, or behind conditional compilation could still legitimately return empty across both substrates. The spec cannot distinguish 'genuinely absent' from 'absent in loaded scope' in those edge cases — documented as a residual, surfaced via projectsSwept so the agent sees how wide the sweep actually was.
- SNIPPET FLAG-THREADING MAY BE INCOMPLETE: threading the project's nullness / treat-warnings-as-errors into check(snippet) fixes the FS3261 case, but FCS snippet compilation does not reproduce every MSBuild-evaluated flag (analyzers, custom targets, source-generator output). snippetFlagsMayDiffer + verdict='unknown' is the safety valve, but over-triggering it would make snippet validation feel useless. Needs tuning against the corpus snippet tasks (Anthropic SDK init-only setters, WithCacheControl overloads).
- LSP DEPRECATE-ALIAS CALLERS MAY NEVER GO IDLE: textDocument_references/definition serve editor/LSP automation that passes raw 0-based coords; their fleet call-rate may stay above the removal threshold indefinitely, leaving the surface at 26 rather than 24. This is acceptable (the contract is preserved) but means the headline 34->24 is a target, not a guarantee — removal is data-gated by design.
- DISPATCHER BECOMES A SINGLE POINT OF FAILURE: routing all 12 old tools + 2 new tools through one Dispatcher.fs means a regression there breaks the entire find/check + legacy surface at once. Mitigated by the Stage 0 mechanical-equivalence DispatcherTests gate (field-superset equality old-vs-new), but the blast radius is larger than today's independent handlers — Stage 0 must ship behind a green full suite before any new capability lands.
- set_project INFRA FAILURES ARE UPSTREAM AND UNFIXED: a failed set_project (e.g. Korat.Cloud Microsoft.VisualStudio.Threading FileNotFoundException) leaves both find and check with a dead context for the whole session. The consolidation does not touch this; both tools should detect a dead context and say so explicitly rather than returning empty/false, but the root cause is out of scope for this spec.

---

_Design produced by a 4-phase multi-agent workflow (evidence → 3-design judge panel → adversarial red-team → spec) over a 41-task corpus mined from this issue. Full spec artifact: `mcp_consolidation_spec.md`. **Decision-ready, NOT yet implemented** — a validation spike of the net-new multi-project `find` sweep precedes any breaking change._

### PR DESCRIPTION
**Pull request by Neftedollar**
Implements **Stage 0 + Stage 1** of the tool-surface consolidation epic (#128): two intent-named tools, `find` and `check`, that hide the FCS/FSAC backend seam and directly attack the agent **rg-fallback** documented across #100 (surface sprawl: 34 tools, ~6k-token routing tax, 22 overlapping-intent pairs).

## What's in this PR

**Stage 0 — dispatcher seam (`refactor`, byte-identical).** The 12 "cluster" handlers route through a new `Dispatcher.fs` (`FindRequest`/`CheckRequest`) instead of calling FcsBridge/LspBridge directly. Concurrency gating + projectPath fallback stay in Program.fs adapters. Proven byte-identical by the existing suite staying green.

**Stage 1 — `find` + `check` (`feat`, additive).**
- **`find(query)`** — multi-project symbol sweep. Unions definitions, references, record-field set sites, and member-usage sites across every member `.fsproj`, recovering cross-project sites single-project `fcs_find_symbol`/`fcs_record_field_audit` structurally miss. **Validated by spike + test: 1→11 sites, 0→5 cross-project** on the fixture reproducing the `TraderRole.Propose` widening that drove agents to `rg`. `matched=false` only when the FCS sweep **and** FSAC index are both empty (no confident cross-project false-negative). Cost: `projectCacheSize` 3→50 + background pre-warm in `set_project`.
- **`check()`** — one trustworthy `verdict` (`clean|errors|unknown`) after a **fresh** in-process type-check; fixes the stale-`{}` false-clean that drove `dotnet build` fallbacks. `speed:"fast"` opts into the cached snapshot; never silently shells out to a build.

**Additive + steered.** The 12 cluster tools are unchanged but their descriptions now steer to `find`/`check`; docs (`tools-detailed.md`, `README`, `AGENT_INTEGRATION.md`) lead with them. Removal of the legacy tools is **telemetry-gated/future**, per the design — not in this PR.

## Verification
- `dotnet restore` — NuGet audit clean.
- `dotnet build -c Release` — 0 warnings (warnings-as-errors).
- **314/314 tests** (303 prior + 4 find + 7 check), confirmed independently.
- `audit-tool-descriptions.py` — green; every description ≤500 chars (largest cluster 497).
- **E2E smoke on the built 0.10.0 binary — 8/8:** 36 tools, `find("FindArgs")` → 30 sites via `fcs-multiproject-sweep`, `find("")` → invalid_args, `check()` → `verdict:clean`.

## Not in scope (deliberate)
Backend-prefix (`[FCS]`/`[FSAC]`) stripping and legacy-tool removal are deferred — they're cross-cutting and telemetry-gated. An A/B protocol (rg-fallback rate, wrong-tool-call rate, tokens-to-first-correct-tool) over the #100 corpus is documented in #128 to measure the win before removing the old tools.

Closes #128.

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: FcsBridge.fs [hunk 1]
```diff
@@ -1977,6 +2103,1124 @@ type internal FcsBridge() =
...
+                                    let fcsLine = line + 1
+                                    let columnsToTry = [| endColumn; startColumn + 1; startColumn |] |> Array.distinct
+
+                                    let symbolUse =
+                                        columnsToTry
+                                        |> Array.tryPick (fun column ->
+                                            checkResults.GetSymbolUseAtLocation(fcsLine, column, lineText, [ text ]))
+
+                                    match symbolUse with
+                                    | Some u -> return Ok u.Symbol.DisplayName
```
[the remark this answers is not in the document]
Comment (Neftedollar) [reply FcsBridge.fs:2215]: Fixed in current main: `kind=position` now keys the subsequent sweep on `Symbol.FullName`, falling back to `DisplayName` only for locals/synthetic symbols with no useful full name — the code comment cites exactly this two-`Config`-namespaces ambiguity.

File: FcsBridge.fs [hunk 2]
```diff
@@ -1977,6 +2103,1124 @@ type internal FcsBridge() =
...
+                not (isNull fullName)
+                && (String.Equals(fullName, query, StringComparison.Ordinal)
+                    || (fullName.EndsWith(query, StringComparison.Ordinal)
+                        && fullName.Length > query.Length
+                        && fullName[fullName.Length - query.Length - 1] = '.'))
+
+            let entityMatchesQuery (e: FSharpEntity) =
+                try
+                    String.Equals(e.DisplayName, query, StringComparison.Ordinal)
+                    || fullNameBoundaryMatch e.FullName
```
[the remark this answers is not in the document]
Comment (Neftedollar) [reply FcsBridge.fs:2344]: Fixed in current main: the declaring-type predicate (`entityMatchesQuery`) now threads `exact` through — `exact=false` does case-insensitive substring matching on both DisplayName and FullName, and the code comment cites the `query="role", field="Propose"` → `TraderRole` case from this thread verbatim.

File: FcsBridge.fs [hunk 3]
```diff
@@ -1977,6 +2103,1124 @@ type internal FcsBridge() =
...
+                        Some
+                            "FSAC has not published diagnostics yet (cold cache); the default trusted check returns a ground-truth verdict."
+                    elif snap.ErrorCount > 0 then
+                        "errors", true, None
+                    else
+                        "clean", true, None
+
+                let nodes =
+                    match snap.ErrorDiagnostics with
+                    | :? JsonArray as arr -> arr |> Seq.map (fun n -> n.DeepClone()) |> Seq.toArray
```
[the remark this answers is not in the document]
Comment (Neftedollar) [reply FcsBridge.fs:2988]: Fixed by the follow-up shipped in PR #130: the FSAC snapshot now retains severity-3/4 entries, and the fast path filters the stored diagnostics by the requested severity floor instead of serving the severity-1-only set.
